File: core/pie.api/src/main/java/mb/pie/api/stamp/resource/Modified.java
package mb.pie.api.stamp.resource;

import mb.resource.ReadableResource;
import mb.resource.hierarchical.HierarchicalResource;
import mb.resource.hierarchical.match.ResourceMatcher;
import mb.resource.hierarchical.walk.ResourceWalker;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.stream.Stream;

class Modified {
    static long modified(ReadableResource resource) throws IOException {
        if(!resource.exists()) {
            return getMaximum();
        } else {
            return modifiedResource(resource);
        }
    }

    static long modified(HierarchicalResource resource, @Nullable ResourceMatcher matcher) throws IOException {
        if(!resource.exists()) {
            return getMaximum();
        } else if(resource.isFile()) {
            return modifiedResource(resource);
        } else if(resource.isDirectory()) {
            return modifiedDir(resource, matcher);
        }
        return getMinimum();
    }

    static long modifiedRec(HierarchicalResource resource, @Nullable ResourceWalker walker, @Nullable ResourceMatcher matcher) throws IOException {
        if(!resource.exists()) {
            return getMaximum();
        } else if(resource.isFile()) {
            return modifiedResource(resource);
        } else if(resource.isDirectory()) {
            return modifiedDirRec(resource, walker, matcher);
        }
        return getMinimum();
    }


    private static long modifiedResource(ReadableResource resource) throws IOException {
        return resource.getLastModifiedTime().toEpochMilli();
    }

    private static long modifiedDir(HierarchicalResource dir, @Nullable ResourceMatcher matcher) throws IOException {
        final long[] lastModified = {getMinimum()}; // Use array to allow access to non-final variable in closure.
        final boolean useMatcher = matcher != null;
        try(final Stream<? extends HierarchicalResource> stream = useMatcher ? dir.list(matcher) : dir.list()) {
            stream.forEach((resource) -> {
                try {
                    final long modified = modifiedResource(resource);
                    lastModified[0] = Math.max(lastModified[0], modified);
                } catch(IOException e) {
                    throw new UncheckedIOException(e);
                }
            });
        } catch(UncheckedIOException e) {
            throw e.getCause();
        }
        return lastModified[0];
    }

    private static long modifiedDirRec(HierarchicalResource dir, @Nullable ResourceWalker walker, @Nullable ResourceMatcher matcher) throws IOException {
        final long[] lastModified = {getMinimum()}; // Use array to allow access to non-final variable in closure.
        final boolean useWalker = walker != null && matcher != null;
        try(final Stream<? extends HierarchicalResource> stream = useWalker ? dir.walk(walker, matcher) : dir.walk()) {
            stream.forEach((resource) -> {
                try {
                    final long modified = modifiedResource(resource);
                    lastModified[0] = Math.max(lastModified[0], modified);
                } catch(IOException e) {
                    throw new UncheckedIOException(e);
                }
            });
        } catch(UncheckedIOException e) {
            throw e.getCause();
        }
        return lastModified[0];
    }

    private static long getMinimum() {
        return Long.MIN_VALUE;
    }

    private static long getMaximum() {
        return Long.MAX_VALUE;
    }
}

File: api/src/main/kotlin/mb/pie/api/stamp/path/Modified.kt
package mb.pie.api.stamp.path

import mb.pie.api.stamp.FileStamp
import mb.pie.api.stamp.FileStamper
import mb.vfs.list.PathMatcher
import mb.vfs.list.PathWalker
import mb.vfs.path.PPath

interface ModifiedFileStamperTrait : FileStamper {
  val unknown get() = Long.MIN_VALUE

  fun modified(path: PPath, matcher: PathMatcher?): Long {
    if(path.isDir) return modifiedDir(path, matcher)
    if(path.isFile) return path.lastModifiedTimeMs()
    return unknown
  }

  fun modifiedRec(path: PPath, walker: PathWalker?): Long {
    if(path.isDir) return modifiedDirRec(path, walker)
    if(path.isFile) return path.lastModifiedTimeMs()
    return unknown
  }

  fun modifiedDir(dir: PPath, matcher: PathMatcher?): Long {
    if(matcher == null) return dir.lastModifiedTimeMs()
    var lastModified = unknown
    for(subPath in matcher.list(dir)) {
      val modified = subPath.lastModifiedTimeMs()
      lastModified = Math.max(lastModified, modified)
    }
    return lastModified
  }

  fun modifiedDirRec(dir: PPath, walker: PathWalker?): Long {
    var lastModified = unknown
    walker?.walk(dir) ?: dir.walk().use { stream ->
      for(subPath in stream) {
        val modified = subPath.lastModifiedTimeMs()
        lastModified = Math.max(lastModified, modified)
      }
    }
    return lastModified
  }
}

data class ModifiedFileStamper(private val matcher: PathMatcher? = null) : ModifiedFileStamperTrait {
  override fun stamp(path: PPath): FileStamp {
    if(!path.exists()) {
      return ValueFileStamp(null, this)
    }
    val modified = modified(path, matcher)
    return ValueFileStamp(modified, this)
  }

  override fun toString(): String {
    return "Modified($matcher)"
  }
}

data class RecModifiedFileStamper(private val walker: PathWalker? = null) : ModifiedFileStamperTrait {
  override fun stamp(path: PPath): FileStamp {
    if(!path.exists()) {
      return ValueFileStamp(null, this)
    }
    val modified = modifiedRec(path, walker)
    return ValueFileStamp(modified, this)
  }

  override fun toString(): String {
    return "RecModified($walker)"
  }
}
